README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

<img src="man/figures/logo.png" align="right" />

# EndTB

<!-- badges: start -->

[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
<!-- badges: end -->

The goal of EndTB is to provide a convenient set of the established TB
models for the simulation, estimation, and projection of Tuberculosis
Epidemic.

Moreover

- the models are implemented in C++ which provides fast simulations.
- the models are called within the [TMB
  framework](https://kaskr.github.io/adcomp/Introduction.html), allowing
  advanced and fast parameter estimation with emprical Bayes approach.

## Installation

You can install the development version of EndTB like so:

``` r
remotes::install_github('kklot/EndTB')
```

Documents of the package and functions can be view as normal R’s
documentation or online at [EndTB.](https://kklot.github.io/EndTB)

## Example

This is a basic example which shows you how to start a model:

``` r
library(EndTB)
#> Welcome to EndTB
TB <- TBM$new(c(sigma = 2))
```

where the default parameters $\sigma$ was replaced with 2.

Steady-state of the model without treatments and no population growth
for that set of parameters can be ploted with.

``` r
TB$plot()
```

<img src="man/figures/README-plot-1.png" width="100%" />

List of states to be plotted can be supplied into the arguments of the
`plot` function above.

Further details can be read in **References** in the top navigation. For
example, [the documentation of
`plot`](https://kklot.github.io/EndTB/reference/TBM.html#method-plot-).

Check out model’s parameters and states for Moldova model.

``` r
EndTB:::states_moldova
#>                        S                      LS1                      AS1 
#>            "Susceptible"         "Sen-1st-Latent"   "Sen-1st-Asymptomatic" 
#>                      IS1                      DS1                      ES1 
#>    "Sen-1st-Symptomatic"      "Sen-1st-Presented"       "Sen-1st-Seekcare" 
#>                      FS1                      LS2                      AS2 
#>          "Sen-1st-Treat"         "Sen-2nd-Latent"   "Sen-2nd-Asymptomatic" 
#>                      IS2                      DS2                      ES2 
#>    "Sen-2nd-Symptomatic"      "Sen-2nd-Presented"       "Sen-2nd-Seekcare" 
#>                      FS2                      RS0                      RS1 
#>          "Sen-2nd-Treat" "Sen-2nd-Recover-Stable"    "Sen-2nd-Relapse-Low" 
#>                      RS2                      LR1                      AR1 
#>   "Sen-2nd-Relapse-High"         "Res-1st-Latent"   "Res-1st-Asymptomatic" 
#>                      IR1                      DR1                      ER1 
#>    "Res-1st-Symptomatic"      "Res-1st-Presented"       "Res-1st-Seekcare" 
#>                      FR1                      LR2                      AR2 
#>          "Res-1st-Treat"         "Res-2nd-Latent"   "Res-2nd-Asymptomatic" 
#>                      IR2                      DR2                      ER2 
#>    "Res-2nd-Symptomatic"      "Res-2nd-Presented"       "Res-2nd-Seekcare" 
#>                      FR2                      RR0                      RR1 
#>          "Res-2nd-Treat" "Res-2nd-Recover-Stable"    "Res-2nd-Relapse-Low" 
#>                      RR2 
#>   "Res-2nd-Relapse-High"
EndTB:::pars_moldova
#>           N      beta_s      beta_r       kappa           b          mu 
#> 2.38428e+07 4.40000e+00 1.20000e+01 1.00000e-01 1.00000e-02 1.50000e-02 
#>       mu_tb     theta_s     theta_r         rho       sigma       delta 
#> 1.50000e-01 1.40000e-01 1.40000e-01 1.00000e-03 7.90000e+00 7.50000e+00 
#>       gamma         phi  varepsilon       omega       tau_0       tau_1 
#> 1.20000e+01 5.20000e+01 9.30000e-01 9.30000e-01 2.00000e+00 5.00000e-01 
#>       chi_s       chi_r      varrho         r_0         r_1         r_2 
#> 2.00000e-01 2.50000e-01 5.00000e-02 3.20000e-02 1.40000e-01 1.50000e-03 
#>         r_3    varsigma        c_s0        c_r0        c_r1         m_n 
#> 7.00000e-01 5.00000e-01 8.00000e-01 3.00000e-01 4.80000e-01 7.40000e-01 
#>         m_r          xi 
#> 4.70000e-01 8.00000e-01
```

## TODO

- [x] 1st revision of the WHO’s model (and currently the only)
- [ ] Revise model to Argentina data
- [ ] Add fitting example
- [ ] Add other model variations


NEWS.md
# EndTB 0.0.0.9000

* First trial version before fixing the compartmental flow


R/class.R
#' Tuberculosis Models
#' 
#' @description TB model
#' @export
TBM <- R6::R6Class("TBM",
list(
    #' @field model TBM in TMB code
    model = NULL,
    #' @field current_pars the parameters used for model simulation
    current_pars = pars_moldova_null,
    #' @description Init the model
    #' @param pars A named vectors of parameters, must matched with the existing parameters (see `pars_moldova` for details)
    #' @param tmax Maximum number of years to simulate
    #' @param dt Time steps, default to 1
    #' @return NULL
    initialize = function(pars = NULL, tmax = 150, dt = 1) {
        if (!is.null(pars)) self$update_pars(pars)
        self$model <- TMB::MakeADFun(
            data = list(init = init_moldova, tmax = tmax, dt = dt),
            parameters = list(pars = self$current_pars),
            DLL = "EndTB", silent = T
        )
    },
    #' @description update model's parameters
    #' @param pars a named vector of parameters 
    update_pars = function(pars) {
        self$current_pars <<- replace(self$current_pars, names(pars), pars)
    },
    #' @description plot model dynamic
    #' @param states list of model states to plot
    #' @return a `ggplot` object.
    plot = function(states = c("S", "AS1", "LS1", "IS1")) {
        requireNamespace("ggplot2", quietly = TRUE)
        p <- private$out_long() %>%
        dplyr::filter(name %in% states) %>%
        ggplot2::ggplot(ggplot2::aes(time, value)) +
        ggplot2::geom_line(ggplot2::aes(color = state)) +
        ggplot2::facet_wrap(~state, nrow = 2, scales = "free") +
        ggplot2::labs(x = "Year", y = "%") +
        ggplot2::theme(legend.position = 'none')
        print(p)
        invisible(p)
    }
),
list(
    out_long = function() {
        self$output %>%
            t() %>%
            dplyr::as_tibble(.name_repair = function(x) make.names(x, unique = TRUE)) %>%
            dplyr::rename_with(~ c("time", names(init_moldova))) %>%
            dplyr::slice(-dplyr::n()) %>%
            dplyr::mutate(dplyr::across(-time, function(x) x / rowSums(dplyr::across(-time)))) %>%
            tidyr::pivot_longer(-time) %>%
            dplyr::mutate(state = factor(name, levels = names(states_moldova), labels = states_moldova))
    }
),
list(
    #' @field output return model's simulation
    output = function() {self$model$report()$out}
)
)

inst/include/TBM.hpp
#include "ode.hpp"

template<class T>
class TB {
  int j = 14;
  double 
  /*0  */ N,
  /*1  */ beta_s,
  /*2  */ beta_r,
  /*3  */ kappa,
  /*4  */ b,
  /*5  */ mu,
  /*6  */ mu_tb,
  /*7  */ theta_s,
  /*8  */ theta_r,
  /*9  */ rho,
  /*10 */ sigma,
  /*11 */ delta,
  /*12 */ gamma,
  /*13 */ phi,
  /*14 */ varepsilon,
  /*15 */ omega,
  /*16 */ tau_0,
  /*17 */ tau_1,
  /*18 */ chi_s,
  /*19 */ chi_r,
  /*20 */ varrho,
  /*21 */ r_0,
  /*22 */ r_1,
  /*23 */ r_2,
  /*24 */ r_3,
  /*25 */ varsigma,
  /*26 */ c_s0,
  /*27 */ c_r0,
  /*28 */ c_r1,
  /*29 */ m_n,
  /*30 */ m_r,
  /*31 */ xi;
public:
  TB(vector<T> pars) :
    N          (asDouble(pars[0])),
    beta_s     (asDouble(pars[1])),
    beta_r     (asDouble(pars[2])),
    kappa      (asDouble(pars[3])),
    b          (asDouble(pars[4])),
    mu         (asDouble(pars[5])),
    mu_tb      (asDouble(pars[6])),
    theta_s    (asDouble(pars[7])),
    theta_r    (asDouble(pars[8])),
    rho        (asDouble(pars[9])),
    sigma      (asDouble(pars[10])),
    delta      (asDouble(pars[11])),
    gamma      (asDouble(pars[12])),
    phi        (asDouble(pars[13])),
    varepsilon (asDouble(pars[14])),
    omega      (asDouble(pars[15])),
    tau_0      (asDouble(pars[16])),
    tau_1      (asDouble(pars[17])),
    chi_s      (asDouble(pars[18])),
    chi_r      (asDouble(pars[19])),
    varrho     (asDouble(pars[20])),
    r_0        (asDouble(pars[21])),
    r_1        (asDouble(pars[22])),
    r_2        (asDouble(pars[23])),
    r_3        (asDouble(pars[24])),
    varsigma   (asDouble(pars[25])),
    c_s0       (asDouble(pars[26])),
    c_r0       (asDouble(pars[27])),
    c_r1       (asDouble(pars[28])),
    m_n        (asDouble(pars[29])),
    m_r        (asDouble(pars[30])),
    xi         (asDouble(pars[31]))
  {}
  void operator()(const state_type &x, state_type &dxdt, const double /* t */) {
    // FOI
    double 
    Nt = std::accumulate(x.begin(), x.end(), 0),
    lambda_s = beta_s / Nt * (x[2] + x[3] + x[7] + x[8] + kappa * (x[4] + x[5] + x[9] + x[10])),
    lambda_r = beta_r / Nt * (x[j+2] + x[j+3] + x[j+7] + x[j+8] + kappa * (x[j+4] + x[j+5] + x[j+9] + x[j+10]));
    // DEs
    dxdt[0]  = b * N - (lambda_r + lambda_s) * x[0] - mu * x[0];
    // Sensitive
    dxdt[1]  = (1 - theta_s) * lambda_s * x[0] - rho * x[1] - mu * x[1];
    dxdt[2]  = theta_s * lambda_s * x[0] + rho * x[1] - sigma * x[2] - mu * x[2] - mu_tb * x[2];
    dxdt[3]  = sigma * x[2] - delta * x[3] - mu * x[3] - mu_tb * x[3];
    dxdt[4]  = delta * x[3] + gamma * x[5] - phi * x[4] - mu * x[4] - mu_tb * x[4];
    dxdt[5]  = (1 - varepsilon * omega) * phi * x[4] - gamma * x[5] - mu * x[5] - mu_tb * x[5];
    dxdt[6]  = varepsilon * omega  * phi * x[4] - (tau_0 + chi_s + varrho) * x[6] - mu * x[6];
    dxdt[7]  = x[12] * r_0 + x[13] * r_1 + x[14] * r_2 - sigma * x[7] - mu * x[7] - mu_tb * x[7];
    dxdt[8]  = sigma * x[7] - delta * x[8] - mu * x[8] - mu_tb * x[8];
    dxdt[9]  = delta * x[8] + gamma * x[10] - phi * x[9] - mu * x[9] - mu_tb * x[9];
    dxdt[10] = (1 - varepsilon * omega) * phi * x[9] - gamma * x[10] - mu * x[10] - mu_tb * x[10];
    dxdt[11] = varepsilon * omega  * phi * x[9] - (tau_0 + chi_s + varrho) * x[11] - mu * x[11];
    dxdt[12] = varsigma * (x[13] + x[14]) - r_0 * x[12] - mu * x[12];
    dxdt[13] =          tau_0 *      c_s0   * (x[6] + x[11]) - varsigma * x[13] - r_1 * x[13] - mu * x[13];
    dxdt[14] = (chi_s + tau_0 * (1 - c_s0)) * (x[6] + x[11]) - varsigma * x[14] - r_2 * x[14] - mu * x[14];
    // MDR
    dxdt[j+1]  = (1 - theta_r) * lambda_r * x[0] - rho * x[j+1] - mu * x[j+1];
    dxdt[j+2]  = theta_r * lambda_r * x[0] + rho * x[j+1] - sigma * x[j+2] - mu * x[j+2] - mu_tb * x[j+2];
    dxdt[j+3]  = sigma * x[j+2] - delta * x[j+3] - mu * x[j+3] -  mu_tb * x[j+3];
    dxdt[j+4]  = delta * x[j+3] + gamma * x[j+5] - phi * x[j+4] - mu * x[j+4] - mu_tb * x[j+4];
    dxdt[j+5]  = (1 - varepsilon * omega) * phi * x[j+4]
               + tau_0 * (1-c_r0) * (1 - xi) * x[j+6] 
               + tau_1 * (1-c_r1) * x[j+15] + chi_r * x[j+15]
               - gamma * x[j+5] - mu * x[j+5] - mu_tb * x[j+5];
    dxdt[j+6]  = varepsilon * omega  * phi * (1 - m_n) * x[j+4] + varrho * x[6] - tau_0 * x[j+6] - mu * x[j+6];
    dxdt[j+7]  = x[j+12] * r_0 + x[j+13] * r_1 + x[j+14] * r_3 - sigma * x[j+7] - mu * x[j+7] - mu_tb * x[j+7];
    dxdt[j+8]  = sigma * x[j+7] - delta * x[j+8] - mu * x[j+8] - mu_tb * x[j+8];
    dxdt[j+9]  = delta * x[j+8] + gamma * x[j+10] - phi * x[j+9] - mu * x[j+9] - mu_tb * x[j+9];
    dxdt[j+10] = (1 - varepsilon * omega) * phi * x[j+9] 
               + tau_0 * (1 - c_r0) * (1 - xi) * x[j+11] 
               + tau_1 * (1 - c_r1) * x[j+16] + chi_r * x[j+16]
               - gamma * x[j+10] - mu * x[j+10] - mu_tb * x[j+10];
    dxdt[j+11] = varepsilon * omega  * phi * (1 - m_r) * x[j+9] + varrho * x[11] - tau_0 * x[j+11] - mu * x[j+11];
    dxdt[j+12] = varsigma * x[j+13] - r_0 * x[j+12] - mu * x[j+12];
    dxdt[j+13] = tau_1 * c_r1 * (x[j+15] + x[j+16]) - varsigma * x[j+13] - r_1 * x[j+13] - mu * x[j+13];
    dxdt[j+14] = tau_0 * c_r0 * (x[j+6] + x[j+11]) - r_3 * x[j+14] - mu * x[j+14];
    dxdt[j+15] = phi * varepsilon * omega * m_n * x[j+4]
               + tau_0 * (1 - c_r0) * xi * x[j+6]
               - (tau_1 + chi_r) * x[j+15] - mu * x[j+15];
    dxdt[j+16] = phi * varepsilon * omega * m_r * x[j+9]
               + tau_0 * (1 - c_r0) * xi * x[j+11]
               - (tau_1 + chi_r) * x[j+16] - mu * x[j+16];
  }
};


inst/include/ode.hpp
#include <boost/numeric/odeint.hpp>
using namespace boost::numeric::odeint;

typedef std::vector<double> state_type;

template <class T, class M> 
struct ODE {
  int n_state, n_time;
  state_type x;
  runge_kutta4<state_type> stepper;
  M model;
  vector<double> track;
  ODE(){};
  ODE(vector<T> init, vector<T> pars, double tmax, double dt) :
    n_state(init.size()), 
    n_time(tmax / dt), 
    x(n_state), 
    model(pars),
    track((n_state + 1) * (n_time + 1))
  {
    // copy starting time
    memcpy(&x[0], &init[0], n_state * sizeof(double));
    *(&track(0)) = 0.0;
    memcpy(&track(0) + 1, &x[0], n_state * sizeof(double));
    // running from time
    int mr = 1;
    for (double t = dt; t < tmax; t += dt) {
      stepper.do_step(model, x, t, dt);
      *(&track(0) + mr * (n_state + 1)) = t;
      memcpy(&track(0) + mr * (n_state + 1) + 1, &x[0], n_state * sizeof(double));
      ++mr;
    }
  };
  matrix<double> out () {
    matrix<double> ans = track.matrix();
    ans.resize(n_state + 1, n_time + 1);
    return ans;
  };
};


R/aaa.R
pars_moldova <- c(
    N          = 23842802, # population
    beta_s     = 4.4,      # Mean rate of transmission per TB case with DS-TB: est. 4.4 (95% CrI 2.4-6)
    beta_r     = 12,       # 
    kappa      = 0.1,      # U(0.1 – 10) Relative infectivity, post vs pre- care-seeking
    b          = 0.01,     # Birth rate: WHO (GHO) (15) – adjusted to yield 1% annual population growth from 1970
    mu         = 0.015,    # Background mortality rate: WHO (GHO)(15), corresponds to mean life expectancy of 68 years
    mu_tb      = 0.15,     # 0,15 (0.14-0.18)
    theta_s    = 0.14,     # Vynnycky E et al. (7)
    theta_r    = 0.14,     # Proportion infections being ‘fast’ progressors to active disease
    rho        = 0.001,    # Horsburgh CR et al. (8) Breakdown to active disease
    sigma      = 7.9,      # est - 7.9 (95% CrI 3.7-11.8) Per-capita hazard of developing symptoms
    delta      = 7.5,      # (95% CrI 2.7-11.5), Per-capita rate of initial presentation to care
    gamma      = 12,       # 12 (95% CrI 8.5-15) Interval between care-seeking episodes
    phi        = 52,       # Treatment initiation delay
    varepsilon = 0.93,     # Probability of diagnosis per patient-provider interaction
    omega      = 0.93,     # Treatment initiation probability, WHO country report 2016-assumption to match TB coverage of 87%, assuming that TB coverage is the product of omega and varepsilon
    tau_0      = 2,        # Treatment duration, first line,
    tau_1      = 0.5,      # Treatment duration, second line,
    chi_s      = 0.2,      # Treatment default rate 
    chi_r      = 0.25,     # Moldova’s National Tb programme 2016-2017
    varrho     = 0.05,     # 0.05 (95% CrI 0.005-0.09) MDR acquisition rate while on FL
    # Relapse, per-capita hazard rates
    r_0        = 0.032,    # relapse following treatment completion
    r_1        = 0.14,     # relapse following treatment default
    r_2        = 0.0015,   # relapse >2 years after treatment 
    r_3        = 0.7,      # relapse in MDR recovered from 1st line
    varsigma   = 0.5,      # ‘Stabilisation’ of relapse risk following treatment, Based in Thomas et al (10): most relapse occurs in first 2 yr after treatment.
    # Treatment success rate (Moldova’s National TB programme 2016-2017)
        c_s0   = 0.8,    # DS-TB in first line treatment, 
        c_r0   = 0.3,      # MDR-TB in first line treatment 
        c_r1   = 0.48,     # MDR-TB in second line treatment
    # Drug sensitivity testing rates (DST) - ECDC 2016 (28)
    m_n        = 0.74,     # new cases
    m_r        = 0.47,     # retreated cases
    xi         = 0.8       # Second line treatment initiation probability
)

pars_moldova_null <- c(
    N          = 23842802, # population
    beta_s     = 4.4,      # Mean rate of transmission per TB case with DS-TB: est. 4.4 (95% CrI 2.4-6)
    beta_r     = 0,        # 
    kappa      = 1,        # U(0.1 – 10) Relative infectivity, post vs pre- care-seeking
    b          = 0.01,     # Birth rate: WHO (GHO) (15) – adjusted to yield 1% annual population growth from 1970
    mu         = 0.01,     # Background mortality rate: WHO (GHO)(15), corresponds to mean life expectancy of 68 years
    mu_tb      = 0.15,     # 0,15 (0.14-0.18)
    theta_s    = 0.14,     # Vynnycky E et al. (7)
    theta_r    = 0.14,     # Proportion infections being ‘fast’ progressors to active disease
    rho        = 0.001,    # Horsburgh CR et al. (8) Breakdown to active disease
    sigma      = 7.9,      # est - 7.9 (95% CrI 3.7-11.8) Per-capita hazard of developing symptoms
    delta      = 0,        # (95% CrI 2.7-11.5), Per-capita rate of initial presentation to care
    gamma      = 0,        # 12 (95% CrI 8.5-15) Interval between care-seeking episodes
    phi        = 0,        # Treatment initiation delay
    varepsilon = 0,        # Probability of diagnosis per patient-provider interaction
    omega      = 0,        # Treatment initiation probability, WHO country report 2016-assumption to match TB coverage of 87%, assuming that TB coverage is the product of omega and varepsilon
    tau_0      = 0,        # Treatment duration, first line,
    tau_1      = 0,        # Treatment duration, second line,
    chi_s      = 0,        # Treatment default rate 
    chi_r      = 0,        # Moldova’s National Tb programme 2016-2017
    varrho     = 0,        # 0.05 (95% CrI 0.005-0.09) MDR acquisition rate while on FL
    # Relapse, per-capita hazard rates
    r_0        = 0,        # relapse following treatment completion
    r_1        = 0,        # relapse following treatment default
    r_2        = 0,        # relapse >2 years after treatment 
    r_3        = 0,        # relapse in MDR recovered from 1st line
    varsigma   = 0,        # ‘Stabilisation’ of relapse risk following treatment, Based in Thomas et al (10): most relapse occurs in first 2 yr after treatment.
    # Treatment success rate (Moldova’s National TB programme 2016-2017)
        c_s0   = 0,        # DS-TB in first line treatment, 
        c_r0   = 0,        # MDR-TB in first line treatment 
        c_r1   = 0,        # MDR-TB in second line treatment
    # Drug sensitivity testing rates (DST) - ECDC 2016 (28)
    m_n        = 0,        # new cases
    m_r        = 0,        # retreated cases
    xi         = 0         # Second line treatment initiation probability
)

init_moldova <- c(
    S   = 23842802,
    LS1 = 1,
    AS1 = 0,
    IS1 = 0,
    DS1 = 0,
    ES1 = 0,
    FS1 = 0,
    LS2 = 0,
    AS2 = 0,
    IS2 = 0,
    DS2 = 0,
    ES2 = 0,
    FS2 = 0,
    RS0 = 0,
    RS1 = 0,
    RS2 = 0,
    # MDR
    LR1 = 0,
    AR1 = 0,
    IR1 = 0,
    DR1 = 0,
    ER1 = 0,
    FR1 = 0,
    LR2 = 0,
    AR2 = 0,
    IR2 = 0,
    DR2 = 0,
    ER2 = 0,
    FR2 = 0,
    RR0 = 0,
    RR1 = 0,
    RR2 = 0,
    G1 = 0,
    G2 = 0
)

states_moldova <- c(
    S   = "Susceptible",
    LS1 = "Sen-1st-Latent",
    AS1 = "Sen-1st-Asymptomatic",
    IS1 = "Sen-1st-Symptomatic",
    DS1 = "Sen-1st-Presented",
    ES1 = "Sen-1st-Seekcare",
    FS1 = "Sen-1st-Treat",
    LS2 = "Sen-2nd-Latent",
    AS2 = "Sen-2nd-Asymptomatic",
    IS2 = "Sen-2nd-Symptomatic",
    DS2 = "Sen-2nd-Presented",
    ES2 = "Sen-2nd-Seekcare",
    FS2 = "Sen-2nd-Treat",
    RS0 = "Sen-2nd-Recover-Stable",
    RS1 = "Sen-2nd-Relapse-Low",
    RS2 = "Sen-2nd-Relapse-High",
    LR1 = "Res-1st-Latent",
    AR1 = "Res-1st-Asymptomatic",
    IR1 = "Res-1st-Symptomatic",
    DR1 = "Res-1st-Presented",
    ER1 = "Res-1st-Seekcare",
    FR1 = "Res-1st-Treat",
    LR2 = "Res-2nd-Latent",
    AR2 = "Res-2nd-Asymptomatic",
    IR2 = "Res-2nd-Symptomatic",
    DR2 = "Res-2nd-Presented",
    ER2 = "Res-2nd-Seekcare",
    FR2 = "Res-2nd-Treat",
    RR0 = "Res-2nd-Recover-Stable",
    RR1 = "Res-2nd-Relapse-Low",
    RR2 = "Res-2nd-Relapse-High"
)